src/Nethermind/Nethermind.BlockValidation/Handlers/ValidateBuilderSubmissionHandler.cs

```diff
@@ -0,0 +1,331 @@
+        }
+        error = null;
+        return true;
+    }
+
+    private bool ValidateBlock(Block block, BidTrace message, long registerGasLimit, out string? error)
+    {
+        error = null;
+
+        if (message.ParentHash != block.Header.ParentHash)
+        {
+            error = $"Parent hash mismatch. Expected {message.ParentHash} but got {block.Header.ParentHash}";
+            return false;
+        }
+
+        if (message.BlockHash != block.Header.Hash)
+        {
+            error = $"Block hash mismatch. Expected {message.BlockHash} but got {block.Header.Hash}";
+            return false;
+        }
+
+        if (message.GasLimit != block.GasLimit)
+        {
+            error = $"Gas limit mismatch. Expected {message.GasLimit} but got {block.GasLimit}";
+            return false;
+        }
+
+        if (message.GasUsed != block.GasUsed)
+        {
+            error = $"Gas used mismatch. Expected {message.GasUsed} but got {block.GasUsed}";
+            return false;
+        }
+
+        Address feeRecipient = message.ProposerFeeRecipient;
+        UInt256 expectedProfit = message.Value;
+
+        if (!ValidatePayload(block, feeRecipient, expectedProfit, registerGasLimit, _blockValidationConfig.UseBalanceDiffProfit, _blockValidationConfig.ExcludeWithdrawals, out error))
+        {
+            return false;
+        }
+
+        _logger.Info($"Validated block Hash: {block.Header.Hash} Number: {block.Header.Number} ParentHash: {block.Header.ParentHash}");
+
+        return true;
+    }
+
+    public Task<ResultWrapper<BlockValidationResult>> ValidateSubmission(BuilderBlockValidationRequest request)
+    {
+        ExecutionPayload payload = request.BlockRequest.ExecutionPayload;
+        BlobsBundleV1 blobsBundle = request.BlockRequest.BlobsBundle;
+
+        string payloadStr = $"BuilderBlock: {payload}";
+
+        _logger.Info($"blobs bundle blobs {blobsBundle.Blobs.Length} commits {blobsBundle.Commitments.Length} proofs {blobsBundle.Proofs.Length}");
+
+        if (!payload.TryGetBlock(out Block? block))
+        {
+            if (_logger.IsWarn) _logger.Warn($"Invalid block. Result of {payloadStr}.");
+            return BlockValidationResult.Invalid($"Block {payload} coud not be parsed as a block");
+        }
+
+        if (block is not null && !ValidateBlock(block, request.BlockRequest.Message, request.RegisterGasLimit, out string? error))
+        {
+            if (_logger.IsWarn) _logger.Warn($"Invalid block. Result of {payloadStr}. Error: {error}");
+            return BlockValidationResult.Invalid(error ?? "Block validation failed");
+        }
+
+        if (block is not null && !ValidateBlobsBundle(block.Transactions, blobsBundle, out string? blobsError))
+        {
+            if (_logger.IsWarn) _logger.Warn($"Invalid blobs bundle. Result of {payloadStr}. Error: {blobsError}");
+            return BlockValidationResult.Invalid(blobsError ?? "Blobs bundle validation failed");
+        }
```

Comment:
A but weird that we do validations after processing a block, shouldn't we do it first? (It would be less costly)

Reply:
https://github.com/flashbots/builder/blob/df9c765067d57ab4b2d0ad39dbb156cbe4965778/eth/block-validation/api.go#L240

I followed the same order as they did, but yeah I agree doing this way would be costly

Reply:
W should consider changing it later, maybe @Ruteri knows why this is the order of checks?